internal/dag/gatewayapi_processor.go

```diff
@@ -1976,6 +1981,81 @@ func (p *GatewayAPIProcessor) httpClusters(routeNamespace string, backendRefs []
 			continue
 		}
 
+		var upstreamValidation *PeerValidationContext
+		var backendRefGroup gatewayapi_v1alpha2.Group
+		if backendRef.Group != nil {
+			backendRefGroup = *backendRef.Group
+		}
+
+		var backendRefKind gatewayapi_v1alpha2.Kind
+		if backendRef.Kind != nil {
+			backendRefKind = *backendRef.Kind
+		}
+
+		var backendNamespace *gatewayapi_v1.Namespace
+		if backendRef.Namespace != nil && *backendRef.Namespace != "" {
+			backendNamespace = backendRef.Namespace
+		} else {
+			backendNamespace = ptr.To(gatewayapi_v1.Namespace(routeNamespace))
+		}
+
+		policyTargetRef := gatewayapi_v1alpha2.PolicyTargetReferenceWithSectionName{
+			PolicyTargetReference: gatewayapi_v1alpha2.PolicyTargetReference{
+				Group:     backendRefGroup,
+				Kind:      backendRefKind,
+				Name:      backendRef.Name,
+				Namespace: backendNamespace,
+			},
+			SectionName: ptr.To(gatewayapi_v1alpha2.SectionName(service.Weighted.ServicePort.Name)),
+		}
+
+		var upstreamTLS *UpstreamTLS
+		// Check to see if there is any BackendTLSPolicy matching this service and service port
+		backendTLSPolicy, found := p.source.LookupBackendTLSPolicyByTargetRef(policyTargetRef)
+		if found {
+			var caSecrets []*Secret
+			for _, certRef := range backendTLSPolicy.Spec.TLS.CACertRefs {
+				if certRef.Group != "" {
+					continue
+				}
+
+				switch certRef.Kind {
+				case "Secret":
+					caSecret, err := p.source.LookupCASecret(types.NamespacedName{
+						Name:      string(certRef.Name),
+						Namespace: backendTLSPolicy.Namespace,
+					}, backendTLSPolicy.Namespace)
+					if err != nil {
+						continue
```

Comment:
we could add some negative unit tests for the rest of these and/or wait for the status for BackendTLSPolicy implementation to cover these errors

might be good at least to add a single test that demonstrates that if there are errors in this area (Policy references a secret that doesn't exist etc.) then there won't be any upstream TLS on the Cluster

Reply:
Added some negative tests, which did happen to catch something wrong in the implementation 😓 , so that is good. When we add conditions the expectations of these tests can probably change to account for that.